geode-core/src/main/java/org/apache/geode/internal/cache/GemFireCacheImpl.java

```diff
@@ -1177,34 +1178,53 @@ public MeterRegistry getMeterRegistry() {
-        logger.info("XML generation completed: {}", cacheXML);
-      } catch (CancelException e) {
-        logger.info("Unable to generate XML description for reconnect of cache due to exception",
-            e);
-      }
-    } else if (sharedConfigEnabled && !getCacheServers().isEmpty()) {
-      // we need to retain a cache-server description if this JVM was started by gfsh
-      List<CacheServerCreation> list = new ArrayList<>(getCacheServers().size());
-      for (Object o : getCacheServers()) {
-        CacheServerImpl cs = (CacheServerImpl) o;
-        if (cs.isDefaultServer()) {
-          CacheServerCreation bsc = new CacheServerCreation(this, cs);
-          list.add(bsc);
+
+  /**
+   * Testable version of saveCacheXmlForReconnect() that allows us to inject an XML generator
+   *
+   * @param xmlGenerator a consumer of a PrintWriter that generates a description of the Cache
+   */
+  protected void prepareForReconnect(Consumer<PrintWriter> xmlGenerator) {
+    boolean sharedConfigEnabled =
+        getInternalDistributedSystem().getConfig().getUseSharedConfiguration();
+
+    try {
+      if (!Boolean.getBoolean(GEMFIRE_PREFIX + "autoReconnect-useCacheXMLFile")
+          && !sharedConfigEnabled) {
+        try {
+          logger.info("generating XML to rebuild the cache after reconnect completes");
+          StringPrintWriter pw = new StringPrintWriter();
+          xmlGenerator.accept(pw);
+          String cacheXML = pw.toString();
+          getCacheConfig().setCacheXMLDescription(cacheXML);
+          logger.info("XML generation completed: {}", cacheXML);
+        } catch (CancelException e) {
+          logger.info("Unable to generate XML description for reconnect of cache due to exception",
+              e);
         }
+      } else if (sharedConfigEnabled && !getCacheServers().isEmpty()) {
+        // we need to retain a cache-server description if this JVM was started by gfsh
+        logger.info("saving cache server configuration for use with the cluster-configuration "
+            + "service on reconnect");
+        List<CacheServerCreation> list = new ArrayList<>(getCacheServers().size());
+        for (Object o : getCacheServers()) {
+          CacheServerImpl cs = (CacheServerImpl) o;
+          if (cs.isDefaultServer()) {
+            CacheServerCreation bsc = new CacheServerCreation(this, cs);
+            list.add(bsc);
+          }
+        }
+        getCacheConfig().setCacheServerCreation(list);
+        logger.info("cache server configuration saved");
       }
-      getCacheConfig().setCacheServerCreation(list);
-      logger.info("CacheServer configuration saved");
+    } catch (Throwable throwable) {
+      logger.info("Saving of cache configuration for auto-reconnect has failed.  "
+          + "Auto-reconnect will be disabled since the cache cannot be rebuilt.", throwable);
+      getInternalDistributedSystem().getConfig().setDisableAutoReconnect(true);
+
```

Comment:
I guess it's ok to not rethrow the `Throwable` here. `saveConfg()` is called from two places in `GMSMembership`:

1. `ManagerImpl.forceDisconnect()`
2. `requestMemberRemoval()`

In both those cases it looks like `setDisableAutoReconnect(true)` will set us on a permanent shutdown path. Since we've logged the `Throwable` here we've fulfilled our obligation to inform the operator.